crates/bevy_ecs/src/query/state.rs

```diff
@@ -40,6 +39,27 @@ impl<Q: WorldQuery, F: WorldQuery> FromWorld for QueryState<Q, F> {
     }
 }
 
+impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
+    pub(crate) fn as_immut(&self) -> &QueryState<Q::ReadOnly, F::ReadOnly> {
+        // Safety: invariant on `WorldQuery` trait upholds that `Q::ReadOnly` and `F::ReadOnly`
+        // have a subset of the access, and match the same archetypes/tables as `Q`/`F` respectively.
+        unsafe { self.as_whatever::<Q::ReadOnly, F::ReadOnly>() }
+    }
+
+    /// # Safety
+    ///
+    /// `NewQ` must have a subset of the access that `Q` does and match the same archetypes/tables
+    /// `NewF` must have a subset of the access that `F` does and match the same archetypes/tables
+    pub(crate) unsafe fn as_whatever<
+        NewQ: WorldQuery<State = Q::State>,
+        NewF: WorldQuery<State = F::State>,
+    >(
+        &self,
+    ) -> &QueryState<NewQ, NewF> {
+        &*(self as *const QueryState<Q, F> as *const QueryState<NewQ, NewF>)
```

Comment:
Does this not rely on the layout of `QueryState` to be identical both generic parameters? I thought we couldn't rely on the internal layout of any `repr(Rust)` type in the general case, or is it different for instantiations of the same generic type?

Reply:
Yea I meant to add `repr(C)` but forgot i guess whoops.